kinto/plugins/accounts/views.py

```diff
@@ -138,3 +223,192 @@ def on_account_changed(event):
+    )
+
+    cache = registry.cache
+    cache.set(cache_key, activation_key, ttl=cache_ttl)
+
+
+def check_validation_key(activation_key, username, registry):
+    """Given a username, compare the activation-key provided with the one from the cache."""
+    hmac_secret = registry.settings["userid_hmac_secret"]
+    cache_key = utils.hmac_digest(hmac_secret, ACCOUNT_VALIDATION_CACHE_KEY.format(username))
+
+    cache = registry.cache
+    cache_result = cache.get(cache_key)
+
+    if cache_result == activation_key:
+        cache.delete(cache_key)  # We're done with the activation key.
+        return True
+    return False
+
+
+@validation.post()
+def post_validation(request):
+    user_id = request.matchdict["user_id"]
+    activation_key = request.matchdict["activation_key"]
+
+    parent_id = user_id
+    try:
+        user = request.registry.storage.get(
+            parent_id=parent_id, resource_name="account", object_id=user_id
+        )
+    except storage_exceptions.ObjectNotFoundError:
+        # Don't give information on the existence of a user id: return a generic error message.
+        error_details = {"message": "Account ID and activation key do not match"}
+        raise http_error(httpexceptions.HTTPForbidden(), **error_details)
+
+    if is_validated(user):
+        error_details = {"message": f"Account {user_id} has already been validated"}
+        raise http_error(httpexceptions.HTTPForbidden(), **error_details)
+
+    if not check_validation_key(activation_key, user_id, request.registry):
+        error_details = {"message": "Account ID and activation key do not match"}
+        raise http_error(httpexceptions.HTTPForbidden(), **error_details)
+
+    # User is now validated.
+    user["validated"] = True
+
+    request.registry.storage.update(
+        parent_id=parent_id, resource_name="account", object_id=user_id, record=user
+    )
+
+    # Send a confirmation email.
+    settings = request.registry.settings
+    email_confirmation_subject_template = settings.get(
+        "account_validation.email_confirmation_subject_template", "Account active"
+    )
+    email_confirmation_body_template = settings.get(
+        "account_validation.email_confirmation_body_template", "The account {id} is now active"
+    )
+    email_sender = settings.get("account_validation.email_sender", "admin@example.com")
+
+    user_email = user["id"]
+    mailer = get_mailer(request)
+    email_context = user.get("email-context", {})
+
+    formatter = MyFormatter()
+    formatted_subject = formatter.format(
+        email_confirmation_subject_template, **user, **email_context
+    )
+    formatted_body = formatter.format(email_confirmation_body_template, **user, **email_context)
+
+    message = Message(
+        subject=formatted_subject,
+        sender=email_sender,
+        recipients=[user_email],
+        body=formatted_body,
+    )
+    mailer.send(message)
```

Comment:
I would suggest to send an event and move all this code to a listener instead (or at least a function)

Reply:
+1 for having a listener for this :-)